File: cpp/Organism.cpp
#include "Organism.h"
#include <iostream>
//Position
Position::Position() { X = -1, Y = -1; }
Position::Position(int xCoordinate, int yCoordinate) :X(xCoordinate), Y(yCoordinate) {}

	//the world is a sphere,
	//so it will go to the other size in the map when it move out of side
Position Position::Left(World* world)			
{
	int x;
	if (X > 0)
		x = X - 1;
	else
		x = world->GetSizeX() - 1;
	return Position(x,Y);
}
Position Position::Right(World* world)
{
	int x;
	if (X < world->GetSizeX() - 1)
		x = X + 1;
	else
		x = 0;
	return Position(x,Y);
}
Position Position::Up(World* world)
{
	int y;
	if (Y > 0)
		y = Y - 1;
	else
		y = world->GetSizeY() - 1;
	return Position(X, y);
}
Position Position::Down(World* world)
{
	int y;
	if (Y < world->GetSizeY() - 1)
		y = Y + 1;
	else
		y = 0;
	return Position(X, y);
}


//constructor
Organism::Organism(char n, int s, int i, int X, int Y, World* world) :
	name(n), strength(s), initiative(i), position(X, Y), world(world) 
{
	if (n == nullOrgan)
		SetActed(true);
}
Organism::Organism(const Organism& o)
{
	SetName(o.GetName());
	SetStrength(o.GetStrength());
	SetInitiative(o.GetInitiative());
	SetPosition(o.GetPosition());
	SetWorld(o.GetWorld());
}


//getters
bool Organism::GetActed()const { return acted; }
int Organism::GetAge()const { return age; }
char Organism::GetName()const { return name; }
int Organism::GetStrength()const { return strength; }
int Organism::GetInitiative()const { return initiative;}
Position Organism::GetPosition()const { return position; }
World* Organism::GetWorld()const { return world; }

//setters
void Organism::SetActed(bool a) { acted = a; }
void Organism::SetAge(int a) { age = a; }
void Organism::SetName(char n) { name = n; }
void Organism::SetStrength(int s) { strength = s; }
void Organism::SetInitiative(int i) { initiative = i; }
void Organism::SetPosition(Position p) { position= p; }
void Organism::SetWorld(World* w) { world = w; }

//other functions
void Organism::Action() {};

void Organism::Collision(Organism*& attacker)
{
	//default collision
	if (attacker->GetStrength() >= this->GetStrength())
	{
		Grid(GetPosition().X, GetPosition().Y) = attacker;

		//if = -1, it is nullorgan or baby of two same animal
		/*if (attacker->GetPosition().X != -1)
			Grid(attacker->GetPosition().X, attacker->GetPosition().Y) = new nullGrid(attacker->GetPosition());*/

		if (attacker->GetPosition().X == -1)
			attacker->SetActed(true);				//same strength, attacker win, so baby win(even if there is something let this be stronger, attacker will not pass code to here)

		attacker->SetPosition(GetPosition());
		delete(this);

		//return false;					//attacker win return false
	}
	else
	{
		//if = -1, it is nullorgan or baby of two same animal
		/*if (attacker->GetPosition().X != -1)
			Grid(attacker->GetPosition().X, attacker->GetPosition().Y) = new nullGrid(attacker->GetPosition());*/

		delete(attacker);

		//return true;					//defender win return true
	}
}

void Organism::Draw() 
{
	std::cout << name <<" ";
}

File: java/src/eti/Project2/Organisms/Organism.java
package eti.Project2.Organisms;

import eti.Project2.MapPosition;
import eti.Project2.World;

import javax.swing.*;
import java.awt.*;
import java.util.Objects;

/**
 * Created By IDEA
 * Author: s188026 Yifei Liu
 * Date: 2021/5/26 22:18
 * Description: animals and plants
 * WOLF        'w'
 * SHEEP       's'
 * FOX         'f'
 * TURTLE      't'
 * ANTELOPE    'a'
 * CYBERSHEEP  'c'
 * GRASS       '!'
 * SOWTHISTLE  '@'
 * GUARANA     '#'
 * BELLADONNA  '$'
 * HOGWEED     '%'
 * HUMAN       'h'
 * nullOrgan   '_'
 */

public class Organism {
    private boolean acted = true;

    private int age = 0;
    private char name;

    private int strength;
    private int initiative;
    private MapPosition position;            //coordinate in the world
    private World world;

    //getters and setters
    public boolean getActed() {
        return acted;
    }

    public void setActed(boolean acted) {
        this.acted = acted;
    }

    public int getAge() {
        return age;
    }

    public void setAge(int age) {
        this.age = age;
    }

    public char getName() {
        return name;
    }

    public void setName(char name) {
        this.name = name;
    }

    public int getStrength() {
        return strength;
    }

    public void setStrength(int strength) {
        this.strength = strength;
    }

    public int getInitiative() {
        return initiative;
    }

    public void setInitiative(int initiative) {
        this.initiative = initiative;
    }

    public MapPosition getPosition() {
        return position;
    }

    public void setPosition(MapPosition position) {
        this.position = position;
    }

    public World getWorld() {
        return world;
    }

    public void setWorld(World world) {
        this.world = world;
    }

    //constructor
    public Organism(char n, int s, int i, int X, int Y, World w) {
        name = n;
        strength = s;
        initiative = i;
        position = new MapPosition(X, Y);
        world = w;
        if (n == '_')
            setActed(true);
    }

    public Organism(Organism o) {
        setName(o.getName());
        setStrength(o.getStrength());
        setInitiative(o.getInitiative());
        setPosition(o.getPosition());
        setWorld(o.getWorld());
        setActed(true);
        setAge(0);
    }

    public Organism() {
    }

    ;

    //other functions
    public void Action() {
    }

    public void Collision(Organism attacker) {
        //default collision
        if (attacker.getStrength() >= this.getStrength()) {
            getWorld().organisms[getPosition().getY()][getPosition().getX()] = attacker;

            if (attacker.getPosition().getX() == -1)
                attacker.setActed(true);                //same strength, attacker win, so baby win(even if there is something let this be stronger, attacker will not pass code to here)

            attacker.setPosition(getPosition());
        }
    }

    public void Draw(JButton b) {
        //std::cout << name <<" ";
        char name = getName();
        ImageIcon img;

        switch (name) {
            case 'a':
                img = (new ImageIcon(Objects.requireNonNull(Organism.class.getResource("PNGImage/antelope.png"))));
                img.setImage(img.getImage().getScaledInstance(40, 40, Image.SCALE_DEFAULT));
                b.setIcon(img);
                break;

            case '$':
                img = (new ImageIcon(Objects.requireNonNull(Organism.class.getResource("PNGImage/belladonna.png"))));
                img.setImage(img.getImage().getScaledInstance(40, 40, Image.SCALE_DEFAULT));
                b.setIcon(img);
                break;
            case 'c':
                img = (new ImageIcon(Objects.requireNonNull(Organism.class.getResource("PNGImage/cybersheep.png"))));
                img.setImage(img.getImage().getScaledInstance(40, 40, Image.SCALE_DEFAULT));
                b.setIcon(img);
                break;
            case 'f':
                img = (new ImageIcon(Objects.requireNonNull(Organism.class.getResource("PNGImage/fox.png"))));
                img.setImage(img.getImage().getScaledInstance(40, 40, Image.SCALE_DEFAULT));
                b.setIcon(img);
                break;
            case '!':
                img = (new ImageIcon(Objects.requireNonNull(Organism.class.getResource("PNGImage/grass.png"))));
                img.setImage(img.getImage().getScaledInstance(40, 40, Image.SCALE_DEFAULT));
                b.setIcon(img);
                break;
            case '#':
                img = (new ImageIcon(Objects.requireNonNull(Organism.class.getResource("PNGImage/guarana.png"))));
                img.setImage(img.getImage().getScaledInstance(40, 40, Image.SCALE_DEFAULT));
                b.setIcon(img);
                break;
            case '%':
                img = (new ImageIcon(Objects.requireNonNull(Organism.class.getResource("PNGImage/hogweed.png"))));
                img.setImage(img.getImage().getScaledInstance(40, 40, Image.SCALE_DEFAULT));
                b.setIcon(img);
                break;
            case 'h':
                img = (new ImageIcon(Objects.requireNonNull(Organism.class.getResource("PNGImage/human.png"))));
                img.setImage(img.getImage().getScaledInstance(40, 40, Image.SCALE_DEFAULT));
                b.setIcon(img);
                break;
            case 's':
                img = (new ImageIcon(Objects.requireNonNull(Organism.class.getResource("PNGImage/sheep.png"))));
                img.setImage(img.getImage().getScaledInstance(40, 40, Image.SCALE_DEFAULT));
                b.setIcon(img);
                break;
            case '@':
                img = (new ImageIcon(Objects.requireNonNull(Organism.class.getResource("PNGImage/sowthistle.png"))));
                img.setImage(img.getImage().getScaledInstance(40, 40, Image.SCALE_DEFAULT));
                b.setIcon(img);
                break;
            case 't':
                img = (new ImageIcon(Objects.requireNonNull(Organism.class.getResource("PNGImage/turtle.png"))));
                img.setImage(img.getImage().getScaledInstance(40, 40, Image.SCALE_DEFAULT));
                b.setIcon(img);
                break;
            case 'w':
                img = (new ImageIcon(Objects.requireNonNull(Organism.class.getResource("PNGImage/wolf.png"))));
                img.setImage(img.getImage().getScaledInstance(40, 40, Image.SCALE_DEFAULT));
                b.setIcon(img);
                break;
            case '_':
                b.setIcon(null);
                break;

        }
    }
}
